### PR TITLE
Generate and ingest curated topic candidate sets (legacy) to SQS [HS-125]

### PR DESCRIPTION
**Pull request by gkatre**
## Goal

Generate and ingest curated topic candidate sets (legacy) to SQS
Benefits:
1. Consolidate the candidate set generation processes in a single flow using the Prefect service (reduces tech debit and code duplication in Metaflow)
2. Integrate Coronavirus (previously a special case) into a single flow with necessary metadata made available as a data model

## Deployment
- [ ] Carolyn has signed off on the changes in candidate sets ([Slack thread](https://pocket.slack.com/archives/C03EYPH93GU/p1659641383324719))

## Implementation Decisions

- Candidate set metadata previously imported and made available as data models are being used
- Curated Candidate set items are also available in the previously built data models making it possible to fetch the necessary data

### Observation:
The `STATIC_TOPIC_MAP_GUIDS` candidate set metadata previously imported from the static json file [topic_map_guids.json](https://github.com/Pocket/dl-metaflow-jobs/blob/main/jobs/resources/topic_map_guids.json#L28-L37) appears to have missing metadata `curatedCandidateID`, `algorithmicCandidateID` and `curatedNotNewTabCandidateID` for the `CORONAVIRUS` topic. For this PR we using the `curatedCandidateID` values.
We would need to update the seed data for the `CORONAVIRUS` topic for the missing values.
(Note: I have applied the [where CURATEDCANDIDATEID is not null and CORPUS_TOPIC_ID is not null](https://github.com/Pocket/data-flows/compare/legacy_topic_flow?expand=1#diff-749c58c9a3c210f4a04a17e86d299b8c58339ce62f6a9c26a1d41a35d1f3e556R32) filter to skip missing data)

## References

JIRA ticket:
* [HS-125](https://getpocket.atlassian.net/browse/HS-125)

### Test run in AWS Dev
The test run results using the AWS Dev account to send to the `RecommendationAPI-Dev` SQS and eventually to DynamoDB table `RECAPI-Dev-CandidateSets` is available [here](https://us-east-1.console.aws.amazon.com/dynamodbv2/home?region=us-east-1#item-explorer?initialTagKey=&table=RECAPI-Dev-CandidateSets)

### CODE REVIEW

File: src/flows/recommendation_api/legacy_topics_flow.py
Comment (jeshuaborges) [inline src/flows/recommendation_api/legacy_topics_flow.py:27]: Im finding understanding what this query very difficult. Perhaps the table names should be more clear and we should add `corpus_topic_id` to the other table so we don't have to join.

Comment (mmiermans) [reply src/flows/recommendation_api/legacy_topics_flow.py:27]: Yeah, the name “topic map guids” is not very descriptive, unless you know there is a file in metaflow named like this. G, maybe you can create a static table with only the data that we need, and choose descriptive names without being ties to how things were named in Metaflow?

File: src/flows/recommendation_api/legacy_topics_flow.py
Comment (mmiermans) [inline src/flows/recommendation_api/legacy_topics_flow.py:28]: This is not a corpus candidate set. Let’s call this column something like “legacy_curated_topic_candidate_set_id”?

File: src/flows/recommendation_api/legacy_topics_flow.py
Comment (mmiermans) [inline src/flows/recommendation_api/legacy_topics_flow.py:62]: This comment applies to the previous line.

**Comment by lfishhead** [comment]
The coronavirus page was distinct from the other topics.  In the old JSON this was indicated by the fact that it was an `editorial_collection` unlike the others.  There are no algorithmic items for it, and you ony need the one candidate set GUID.

My attitude is that most of the things that are unclear or clumsy here stem from decisions made in the creation of the recommendation-api.  It's hard to undo a lot of that, esp. when migrating candidate sets that are to be deprecated eventually?  I am curious why we aren't making the legacy slates pull corpus candidate sets, or how we anticipate doing that work later.

**Comment by mmiermans** [comment]
> I am curious why we aren't making the legacy slates pull corpus candidate sets, or how we anticipate doing that work later.

The legacy slates are required to return a `resolved_id`. We could've decided to create hybrid candidate sets which identify both the CorpusItem and the `resolved_id`. Considering that we expect all legacy slates to be phased out, I think it's cleaner and easier to duplicate the flows and write to SQS, instead of having to change legacy slates to fetch candidates from the Feature Group.

File: src/flows/recommendation_api/legacy_topics_flow.py
Comment (gkatre) [reply src/flows/recommendation_api/legacy_topics_flow.py:27]: Yes. We can create a new static table in Snowflake dbt with better naming and just the 2 needed columns. We also need to include the missing metadata for the `CORONAVIRUS` topic in the table

File: src/flows/recommendation_api/legacy_topics_flow.py
Comment (gkatre) [reply src/flows/recommendation_api/legacy_topics_flow.py:28]: > This is not a corpus candidate set. Let’s call this column something like “legacy_curated_topic_candidate_set_id”?

We will have to do this in a separate Snowflake dbt PR. This is a Prefect PR that ingests the candidate set to SQS

**Comment by mmiermans** [comment]
Let's add a deployment check box in the PR description to get sign-off from the editorial team:

- [ ] Carolyn has signed off on the changes in candidate sets ([Slack thread](https://pocket.slack.com/archives/C03EYPH93GU/p1659641383324719))

**Comment by mmiermans** [comment]
Coronavirus candidate set id: c29fbf48-0093-49d2-ae55-4e01a3cf800e

**Comment by lfishhead** [comment]
> Considering that we expect all legacy slates to be phased out.....
my question is when.  my concern is that we will maintain parallel versions of the same slates for a long time.

**Comment by gkatre** [comment]
Updated the flow using the new static metadata table for legacy topic candidate set IDs

**Comment by gkatre** [comment]
Based on the discussion with Matt, updated the query to limit the candidates per topic to 45.

@lfishhead , @mmiermans 
When you get a moment could both do a final review and approve for merge if things look ok?

File: src/common_tasks/corpus_candidate_set.py
Comment (lfishhead) [inline src/common_tasks/corpus_candidate_set.py:28]: we may wish to increase this to 3 if we expect a full slate.

**Review by lfishhead — APPROVED** [review]
thanks G!

File: src/common_tasks/corpus_candidate_set.py
Comment (gkatre) [reply src/common_tasks/corpus_candidate_set.py:28]: Those are tests to verify the minimum state of the topic candidates and causing the flow to fail if those conditions are not met. 
If we are to increase the `min_item_count` to 3, we would have to apply filters to not have those topics when the candidates are less than 3.  Is that what you want?

**Review by mmiermans — APPROVED** [review]
:+1: lgtm

File: src/common_tasks/corpus_candidate_set.py
Comment (gkatre) [reply src/common_tasks/corpus_candidate_set.py:28]: @lfishhead I have merged the changes for now. If the candidates per topic min count 3 is important, lets discuss it and make the change in the next iteration

**Pull request merged**
